Release/src/http/client/http_client_asio.cpp

```diff
@@ -363,102 +375,331 @@ class asio_context : public request_context, public std::enable_shared_from_this
+                std::string http_version;
+                response_stream >> http_version;
+                status_code status_code;
+                response_stream >> status_code;
+                
+                if (!response_stream || http_version.substr(0, 5) != "HTTP/")
+                {
+                    m_context->report_error("Invalid HTTP status line during proxy connection", ec, httpclient_errorcode_context::readheader);
+                    return;
+                }
+                
+                if (status_code != 200)
+                {
+                    utility::stringstream_t err_ss;
+                    err_ss << U("Expected a 200 response from proxy, received: ") << status_code;
+                    m_context->report_error(err_ss.str(), ec, httpclient_errorcode_context::readheader);
+                    return;
+                }
+                
+                m_context->m_connection->upgrade_to_ssl();
+                
+                m_ssl_tunnel_established(m_context);
+            }
+            else
+            {
+                // These errors tell if connection was closed.
+                const bool socket_was_closed((boost::asio::error::eof == ec)
+                    || (boost::asio::error::connection_reset == ec)
+                    || (boost::asio::error::connection_aborted == ec));
+                if (socket_was_closed && m_context->m_connection->is_reused())
+                {
+                    // Failed to write to socket because connection was already closed while it was in the pool.
+                    // close() here ensures socket is closed in a robust way and prevents the connection from being put to the pool again.
+                    m_context->m_connection->close();
+            
+                    // Create a new context and copy the request object, completion event and
+                    // cancellation registration to maintain the old state.
+                    // This also obtains a new connection from pool.
+                    auto new_ctx = m_context->create_request_context(m_context->m_http_client, m_context->m_request);
+                    new_ctx->m_request_completion = m_context->m_request_completion;
+                    new_ctx->m_cancellationRegistration = m_context->m_cancellationRegistration;
+            
+                    auto client = std::static_pointer_cast<asio_client>(m_context->m_http_client);
+                    // Resend the request using the new context.
+                    client->send_request(new_ctx);
+                }
+                else
+                {
+                    m_context->report_error("Failed to read HTTP status line from proxy", ec, httpclient_errorcode_context::readheader);
+                }
+            }
         }
+    
+        std::function<void(std::shared_ptr<asio_context>)> m_ssl_tunnel_established;
+        std::shared_ptr<asio_context> m_context;
+    
+        boost::asio::streambuf request_;
```

Comment:
As per our coding conventions, we try to follow m_ prefix for naming class members. It will be good to be consistent.